### ISSUE TITLE
GPU is slow when parallel_shots is many 

### ISSUE THREAD

**Issue by hhorii** [Qiskit/qiskit-aer#793]
### Informations

- **Qiskit Aer version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?

`statevector_gpu` is very slow if `parallel_shots` is many.

### Steps to reproduce the problem
Code: 
```
from qiskit import execute
from qiskit.circuit.library import QFT
from qiskit.providers.aer import QasmSimulator
from qiskit.providers.aer.noise import NoiseModel, amplitude_damping_error

simulator = QasmSimulator()


noise_model = NoiseModel()
noise_model.add_all_qubit_quantum_error(amplitude_damping_error(1e-3), ['h'])

qubits = 10
qft = QFT(qubits).inverse()
for q in range(qubits):
    qft.h(q)

qft.measure_all()

backend_options = {}
backend_options['method'] = 'statevector'
result = execute(qft, simulator, noise_model=noise_model, backend_options=backend_options).result()
print("statevector: " + str(result.to_dict()['time_taken']))


backend_options['method'] = 'statevector_gpu'
result = execute(qft, simulator, noise_model=noise_model, backend_options=backend_options).result()
print("statevector_gpu: " + str(result.to_dict()['time_taken']))
```

Result:
```
statevector: 0.06392407417297363
statevector_gpu: 18.052891492843628
```

### What is the expected behavior?

GPU is faster than CPU

### Suggested solutions

Current thrust has many synchronization between CPU and GPU.
Running multiple `QubitVectorThrust` is not efficient.

A work around is to generate a long qubit vector that consists of multiple qubit vectors.

**Comment by chriseclectic** [comment]
I think our current shot/circuit parallelism paradigm in general will work very poorly for GPU simulator because the outerloops are all handled by OMP threads that spawn multiple instances of the state classes.

**Comment by doichanj** [comment]
Multi-shots on GPU simulator get better, but still behind CPU

`statevector: 1.1290619373321533
statevector_gpu: 1.4679443836212158`

using the latest code https://github.com/Qiskit/qiskit-aer/pull/788
(on P9 + 6x Tesla V100)

**Comment by doichanj** [comment]
I optimized multi-shot parallelization with GPUs in the latest version of GPU + MPI simulator (it is not on Github yet)

The result with Hiroshi's test code 
`statevector: 1.13724946975708
statevector_gpu: 0.310870885848999`
(on Power9 with 6 Tesla V100, using 80 threads (2 SMT/core))

I would like to put the optimized codes separately from #788 , and I will make another pull request to solve this issue after #788 will be merged to the master.

**Comment by hhorii** [comment]
In general, if memory is enough to run density matrix, density matrix is better than statevector for noise simulation. I think that most of GPU machines are workstation with much memory (such as 128GB-) and can simulate 16-qubit with density matrix (64GB with double precision). Therefore, multi-shot-parallel is necessary for 17 or more qubits for noise simulation.

If problems do not exist in multi-shot-parall simulation for 17 qubits, this issue can be closed by enabling `Method::automatic` to configure`Method::density_matrix_thrust_gpu`. 
https://github.com/Qiskit/qiskit-aer/blob/7e6ed614cc87156414061bcbfb1c3c1cc8be6b24/src/controllers/qasm_controller.hpp#L783-L789

An issue in multi-experiment-parallel simulation for small qubits will remain. Maybe, this issue is a problem to simulate parameterized circuits. I believe that it should be another issue because other techniques are available to optimize simulation of these circuits.

**Comment by hhorii** [comment]
This issue was resolved with batched execution in 0.10.
